lib/fastlane/plugin/firebase_app_distribution/actions/firebase_app_distribution_action.rb

```diff
@@ -183,27 +211,50 @@ def self.get_token
 
       def self.validate_app!(app_id)
         token = get_token
-
-        connection = Faraday.new(url: BASE_URL) do |conn|
-          conn.response(:json, parser_options: { symbolize_names: true })
-          conn.response(:raise_error) # raise_error middleware will run before the json middleware
-          conn.adapter(Faraday.default_adapter)
-        end
         begin
-          response = connection.get("#{PATH}#{app_id}") do |request|
+          response = CONNECTION.get("#{PATH}#{app_id}") do |request|
             request.headers["Authorization"] = "Bearer " + token
           end
         rescue Faraday::ResourceNotFound => error
           UI.user_error!("App Distribution could not find your app #{app_id}. Make sure to onboard your app by pressing the \"Get started\" button on the App Distribution page in the Firebase console: https://console.firebase.google.com/project/_/appdistribution")
         rescue => error
           UI.crash!("Failed to fetch app information: #{error.message}")
         end
-
         contact_email = response.body[:contactEmail]
 
         if contact_email.strip.empty?
           UI.user_error!("We could not find a contact email for app #{app_id}. Please visit App Distribution within the Firebase Console to set one up.")
         end
+        return "projects/#{response.body[:projectNumber]}/apps/#{response.body[:appId]}/releases/-/binaries/"
+      end
+
+      def self.upload_binary!(app_id, binary_path)
+        token = get_token
+        begin
+          response = CONNECTION.post("/app-binary-uploads?app_id=#{app_id}", File.open(binary_path).read) do |request|
+            request.headers["Authorization"] = "Bearer " + token
+            # request.headers["X-APP-DISTRO-API-CLIENT-ID"] = "FastLane"
+          end
+          response.status
+        rescue Faraday::ResourceNotFound => error
+          UI.user_error!("App Distribution could not find your app #{app_id}. Make sure to onboard your app by pressing the \"Get started\" button on the App Distribution page in the Firebase console: https://console.firebase.google.com/project/_/appdistribution")
+        rescue => error
+          UI.crash!("Failed to upload distribution: #{error.message}")
+        end
+      end
+
+      def self.upload_status!(app_token, app_id, binary_path)
+        token = get_token
+        upload_token = CGI.escape(app_token + Digest::SHA256.hexdigest(File.open(binary_path).read))
```

Comment:
Since you call `upload_status!` once per iteration while polling, it would be good to avoid re-reading and hashing the entire binary each time through the loop.